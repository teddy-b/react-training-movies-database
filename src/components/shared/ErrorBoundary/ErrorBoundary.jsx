/* @flow */

import * as React from 'react'

type Props = {
  children: ?React.Node,
}

type State = {
  error: boolean,
  errorInfo: {
    componentStack: string
  }
}

class ErrorBoundary extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      error: false,
      errorInfo: {
        componentStack: ''
      }
    }
  }

  componentDidCatch(error: boolean, errorInfo: { componentStack: string }) {
    this.setState({
      error,
      errorInfo
    })
  }

  render(): React.Node {
    if (this.state.error) {
      return (
        <div>
          <h2>Something went wrong.</h2>
          <details style={{
            whiteSpace: 'pre-wrap'
          }}
          >
            {this.state.error && this.state.error.toString()}
            <br />
            {this.state.errorInfo.componentStack}
          </details>
        </div>
      )
    }

    return this.props.children
  }
}

export default ErrorBoundary
