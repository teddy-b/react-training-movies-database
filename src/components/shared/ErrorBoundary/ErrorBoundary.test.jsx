import React from 'react'

import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

import ErrorBoundary from './ErrorBoundary'

describe('ErrorBoundary', () => {
  it('renders correctly without errors', () => {
    const SomeComponent = () => (
      <div>Some data</div>
    )
    const component = renderer.create(
      <ErrorBoundary>
        <SomeComponent />
      </ErrorBoundary>
    )
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should catch errors with componentDidCatch', () => {
    const SomeComponent = () => {
      throw new Error('Error thrown from problem child')
    }

    jest.spyOn(ErrorBoundary.prototype, 'componentDidCatch')

    mount(
      <ErrorBoundary>
        <SomeComponent />
      </ErrorBoundary>
    )

    expect(ErrorBoundary.prototype.componentDidCatch).toHaveBeenCalled()
  })
})
