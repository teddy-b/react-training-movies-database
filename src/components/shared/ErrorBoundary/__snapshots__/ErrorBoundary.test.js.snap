// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary renders correctly without errors 1`] = `
<div>
  Some data
</div>
`;
